Stepper: Add Form Integration How To topic (#7464)
DevExpress/devextreme-documentation: concepts/05 UI Components/Stepper/30 Implement a Wizard/10 Synchronize Steps and Navigation Panel.md

        // ...
        onSelectionChanged(e) {
            const selectedIndex = e.component.option('selectedIndex');

            setSelectedIndex(selectedIndex);
        },
    }).dxStepper('instance');

    const prevButton = $('#prevButton').dxButton({
        // ...
        onClick: () => {
            const selectedIndex = stepper.option('selectedIndex');

            setSelectedIndex(selectedIndex - 1);
        },
        visible: false,
    }).dxButton('instance');

    const nextButton = $('#nextButton').dxButton({
        // ...
        onClick: () => {
            const selectedIndex = stepper.option('selectedIndex');

            if (selectedIndex < steps.length - 1) {
                setSelectedIndex(selectedIndex + 1);
            }
        },
    }).dxButton('instance');

    function setSelectedIndex(index) {
        stepper.option('selectedIndex', index);
        stepContent.option('selectedIndex', index);
        setCurrentStepCaption(index);
        updateStepNavigationButtons(index);
        
        if (index === steps.length - 1) {
            stepContent.option('items[4].template', getConfirmationTemplate());
        }
    }

> `setSelectedIndex` also updates the navigation panel with `updateStepNavigationButtons` and `setCurrentStepCaption`.

Reviewer: ```suggestion
`setSelectedIndex` also updates the navigation panel. It calls the following methods: `updateStepNavigationButtons` and `setCurrentStepCaption`.
```

> On the last step, the "Next" button is changed to "Confirm". The "Confirm" button submits the form and disables user interactions with Stepper. To learn how to disable Stepper interactions, refer to the [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/) help topic.

Reviewer: ```suggestion
On the last step, the "Next" button is changes to "Confirm". The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

> On the last step, the "Next" button is changes to "Confirm". The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).

Reviewer 2: ```suggestion
On the last step, the "Next" button changes to "Confirm". The "Confirm" button submits the form and disables user interactions with the Stepper. For instructions on how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

            $('.current-step').append(`Step <span class="selected-index">${index + 1}</span> of ${steps.length}`);
        } else {
            $('.selected-index').text(index + 1);
        }
    }

    function reset() {
        confirmed = false;
        setSelectedIndex(0);
        // ...
    }

    function confirm() {
        confirmed = true;
        setSelectedIndex(steps.length - 1);
        // ...
    }

    function updateStepNavigationButtons(index) {
        const isLastStep = index === steps.length - 1;
        const lastStepNextButtonText = confirmed ? 'Reset' : 'Confirm';
        const nextButtonText = isLastStep ? lastStepNextButtonText : 'Next';

        prevButton.option('visible', !!index && !confirmed);
        nextButton.option('text', nextButtonText);
    }

    function setStepperReadonly(readonly) {
        stepper.option('focusStateEnabled', !readonly);
        
        if (readonly) {
            stepper.option('elementAttr', { class: 'readonly' });
        } else {
            stepper.resetOption('elementAttr');
        }
    }

    <!-- tab: styles.css -->
    .readonly {
        pointer-events: none;
    }

##### Angular

> To synchronize the **selectedIndex** properties of Stepper and MultiView, assign them to a common variable. Integrate two-way data binding syntax to synchronize Stepper [selectedIndex](/Documentation/ApiReference/UI_Components/dxStepper/Configuration/#selectedIndex) with the common variable:

Reviewer: I'm not sure what's going on, but the code shows the same syntax for the Stepper and MultiView. I guess they both use two-way data binding? 

```suggestion
Use a separate variable and two-way data binding syntax to synchronize **selectedIndex** properties of Stepper and MultiView:
```

Author: Yeah they both use two-way binding in the accompanying Form Integration demo, though it is only necessary on the Stepper component.

    <!-- tab: app.component.html -->
    <dx-stepper ...
        [(selectedIndex)]="selectedIndex"
    >
    </dx-stepper>
    <div class="content">
        <dx-multi-view ...
            [(selectedIndex)]="selectedIndex"
        >
        </dx-multi-view>
    </div>

    <!-- tab: app.component.ts -->
    // ...
    export class AppComponent {
        selectedIndex: number;

        constructor(private readonly appService: AppService) {
            // ...
            this.selectedIndex = 0;
        }
    }

> This example utilizes the `selectedIndex` variable in multiple places to implement navigation panel functionality:

Reviewer: ```suggestion
This example utilizes the `selectedIndex` variable to implement navigation panel functionality:
```

- The `onPrevButtonClick` and `onNextButtonClick` handlers move to the previous/next step by modifying `selectedIndex`.
- The `getNextButtonText` function changes the text of the "Next" button to "Confirm" on the last step.

> The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).

Reviewer 2: ```suggestion
The "Confirm" button submits the form and disables user interactions with the Stepper. For details on how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

> The "Confirm" button submits the form and disables user interactions with Stepper. To learn how to disable Stepper interactions, refer to the [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/) help topic.

Reviewer: ```suggestion
The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

        constructor(private readonly appService: AppService) {
            // ...
            this.isConfirmed = false;
            this.isStepperReadonly = false;
        }

        getNextButtonText() {
            if (this.selectedIndex < this.steps.length - 1) {
                return 'Next';
            }

            return this.isConfirmed ? 'Reset' : 'Confirm';
        }

        onPrevButtonClick() {
            this.selectedIndex -= 1;
        }

        moveNext() {
            this.selectedIndex += 1;
        }

        reset(){
            this.isConfirmed = false;
            this.selectedIndex = 0;
            this.isStepperReadonly = false;
            // ...
        }

        confirm(){
            this.isConfirmed = true;
            this.isStepperReadonly = true;
        }

        onNextButtonClick() {
            if (this.selectedIndex < this.steps.length - 1) {
                this.moveNext();
            } else if (this.isConfirmed) {
                this.reset();
            } else {
                this.confirm();
            }
        }
    }

    <!-- tab: app.component.css -->
    ::ng-deep .readonly {
        pointer-events: none;
    }

##### Vue

> To synchronize the **selectedIndex** properties of Stepper and MultiView, assign them to a common variable. Integrate the `v-model` directive to synchronize Stepper [selectedIndex](/Documentation/ApiReference/UI_Components/dxStepper/Configuration/#selectedIndex) with the common variable:

Reviewer: ```suggestion
Use a separate variable and the `v-model` directive to synchronize **selectedIndex** properties of Stepper and MultiView:
```

    <!-- tab: App.vue -->
    <template>
        <DxStepper ...
            v-model:selected-index="selectedIndex"
        >
            <DxStepperItem ... />
        </DxStepper>
        <div class="content">
            <DxMultiView ...
                v-model:selected-index="selectedIndex"
            >
            </DxMultiView>
        </div>
    </template>

    <script setup lang="ts">
    const selectedIndex = ref(0);

    // ...
    </script>

> This example utilizes the `selectedIndex` variable in multiple places to implement navigation panel functionality:

Reviewer: ```suggestion
This example utilizes the `selectedIndex` variable to implement navigation panel functionality:
```

- The `onPrevButtonClick` and `onNextButtonClick` handlers move to the previous/next step by modifying `selectedIndex`.
- The `nextButtonText` function changes the text of the "Next" button to "Confirm" on the last step.

> The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).

Reviewer 2: ```suggestion
The "Confirm" button submits the form and disables user interactions with the Stepper. For details on how to disable Stepper interactions, refer to the following help topic [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

> The "Confirm" button submits the form and disables user interactions with Stepper. To learn how to disable Stepper interactions, refer to the [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/) help topic.

Reviewer: ```suggestion
The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

                    :text="nextButtonText"
                    @click="onNextButtonClick"
                />
            </div>
        </div>
    </template>

    <script setup lang="ts">
    // ...
    const isConfirmed = ref(false);
    const isStepperReadonly = ref(false);

    const nextButtonText = computed(() => {
        if (selectedIndex.value < steps.value.length - 1) {
            return 'Next';
        }

        return isConfirmed.value ? 'Reset' : 'Confirm';
    });

    function onPrevButtonClick() {
        selectedIndex.value -= 1;
    }

    const moveNext = () => {
        selectedIndex.value += 1;
    };

    const reset = () => {
        isConfirmed.value = false;
        selectedIndex.value = 0;
        isStepperReadonly.value = false;
        // ...
    };

    const confirm = () => {
        isConfirmed.value = true;
        isStepperReadonly.value = true;
    };

    function onNextButtonClick() {
        if (selectedIndex.value < steps.value.length - 1) {
            moveNext();
        } else if (isConfirmed.value) {
            reset();
        } else {
            confirm();
        }
    }
    </script>

##### React

> To synchronize the **selectedIndex** properties of Stepper and MultiView, update the [selectedIndex](/Documentation/ApiReference/UI_Components/dxMultiView/Configuration/#selectedIndex) property of MultiView in the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxStepper/Configuration/#onSelectionChanged) handler of Stepper.

Reviewer: ```suggestion
To synchronize the **selectedIndex** properties of Stepper and MultiView, update the MultiView.[selectedIndex](/Documentation/ApiReference/UI_Components/dxMultiView/Configuration/#selectedIndex) property in the Stepper.[onSelectionChanged](/Documentation/ApiReference/UI_Components/dxStepper/Configuration/#onSelectionChanged) event handler.
```

    <!-- tab: App.tsx -->
    // ...
    export default function App () {
        const [selectedIndex, setSelectedIndex] = useState(0);

        const onSelectionChanged = useCallback(({ component }: SelectionChangedEvent) => {
            setSelectedIndex(component.option('selectedIndex') ?? 0);
        }, []);

        // ...
        return (
            <>
                <Stepper ...
                    selectedIndex={selectedIndex}
                    onSelectionChanged={onSelectionChanged}
                >
                </Stepper>
                <div className="content">
                    <MultiView ...
                        selectedIndex={selectedIndex}
                    >
                    </MultiView>
                </div>
            </>
        )
    }

> This example utilizes the `selectedIndex` variable in multiple places to implement navigation panel functionality:

Reviewer: ```suggestion
This example utilizes the `selectedIndex` variable to implement navigation panel functionality:
```

- The `onPrevButtonClick` and `onNextButtonClick` handlers move to the previous/next step by modifying `selectedIndex`.
- The `nextButtonText` function changes the text of the "Next" button to "Confirm" on the last step.

> The "Confirm" button submits the form and disables user interactions with Stepper. To learn how to disable Stepper interactions, refer to the [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/) help topic.

Reviewer: ```suggestion
The "Confirm" button submits the form and disables user interactions with the Stepper. To learn how to disable Stepper interactions, refer to the following help topic: [Configure a Readonly Stepper](/Documentation/Guide/UI_Components/Stepper/Configure_a_Read-Only_Stepper/).
```

---

##### jQuery

> To synchronize the **selectedIndex** properties of Stepper and MultiView, create a new function, `setSelectedIndex`. Call it in the [onSelectionChanged](/Documentation/ApiReference/UI_Components/dxStepper/Configuration/#onSelectionChanged) handler of Stepper and the [onClick]() handlers of both navigation buttons:

Reviewer 3: It is difficult to figure out what is meant. I would say smth like (just in better words)
To synchronize selected Stepper step and MultiView content we pass same value as selectedIndex properties. We set this property in `onSelectionChanged` to support change by step click, and to buttons - to support wizard buttons

            setSelectedIndex(selectedIndex);
        },
    }).dxStepper('instance');

    const prevButton = $('#prevButton').dxButton({
        // ...
        onClick: () => {
            const selectedIndex = stepper.option('selectedIndex');

            setSelectedIndex(selectedIndex - 1);
        },
        visible: false,
    }).dxButton('instance');

    const nextButton = $('#nextButton').dxButton({
        // ...
        onClick: () => {
            const selectedIndex = stepper.option('selectedIndex');

            if (selectedIndex < steps.length - 1) {
                setSelectedIndex(selectedIndex + 1);
            }
        },
    }).dxButton('instance');

    function setSelectedIndex(index) {
        stepper.option('selectedIndex', index);
        stepContent.option('selectedIndex', index);
        setCurrentStepCaption(index);
        updateStepNavigationButtons(index);
        
        if (index === steps.length - 1) {
            stepContent.option('items[4].template', getConfirmationTemplate());
        }
    }

`setSelectedIndex` also updates the navigation panel with `updateStepNavigationButtons` and `setCurrentStepCaption`.

> `setCurrentStepCaption` updates the active step indicator as users progress through the Stepper:

Reviewer 3: not sure it is important to emphasize those functions - it is not about Stepper, we can say that we also update wizard caption and button states (names, visibility)

    <!-- tab: index.js -->
    let confirmed = false;

    function setCurrentStepCaption(index) {
        if (confirmed) {
            $('.current-step').empty();
        } else if (!$('.current-step').text()) {
            $('.current-step').append(`Step <span class="selected-index">${index + 1}</span> of ${steps.length}`);
        } else {
            $('.selected-index').text(index + 1);
        }
    }

> `updateStepNavigationButtons` displays the "Back" button on the second step. On the last step, the "Next" button changes to "Confirm". When clicked, the "Confirm" button does the following:

Reviewer 3: imho, to many information about side function

> - Submits the form.

Reviewer 3: Looks like it should be below, before `confirm` function code

        constructor(private readonly appService: AppService) {
            // ...
            this.isConfirmed = false;
            this.isStepperReadonly = false;
        }

        getNextButtonText() {
            if (this.selectedIndex < this.steps.length - 1) {
                return 'Next';
            }

            return this.isConfirmed ? 'Reset' : 'Confirm';
        }

        onPrevButtonClick() {
            this.selectedIndex -= 1;
        }

        moveNext() {
            this.selectedIndex += 1;
        }

        reset(){
            this.isConfirmed = false;
            this.selectedIndex = 0;
            this.isStepperReadonly = false;
            // ...
        }

        confirm(){
            this.isConfirmed = true;
            this.isStepperReadonly = true;
        }

        onNextButtonClick() {
            if (this.selectedIndex < this.steps.length - 1) {
                this.moveNext();
            } else if (this.isConfirmed) {
                this.reset();
            } else {
                this.confirm();
            }
        }
    }

    <!-- tab: app.component.css -->
    ::ng-deep .readonly {
        pointer-events: none;
    }

##### Vue

> To synchronize the **selectedIndex** properties of Stepper and MultiView, assign them to a common variable using the `v-model` directive:

Reviewer 3: We use `v-model` not to sync 2 components, but to allow 2-way binding for the property. Not sure it is required for MultiView, though. We sync it the same way as in other frameworks - by setting same value and add some logic to change this value. `v-model` allows to change this value on step click in Stepper
